### ISSUE TITLE
Could `no-undef` allow unknown globals if they are inside an If statement that checks for them?

### ISSUE THREAD

**Issue by Mr0grog** [eslint/eslint#7636]
* **ESLint Version:** 3.10.2
* **Node Version:** 6.9.1
* **npm Version:** 3.10.8
* **Parser:** Default

I often see code that checks for the presence of a global variable before using it when dealing with things only available on certain platforms features only available in newer engines. For example, in a codebase I’m currently working on:

```js
if (window.WebKitPoint) {
  const zeroPoint = new WebKitPoint(0, 0);
  // do some stuff with this point
}
else {
  // some less-great solution to the task
}
```

The `no-undef` rule quite reasonably flags this as an undefined global. While one way to solve this is to manually list those globals in an ESLint config file or at the top of the file, I wonder if it would be helpful or useful to not flag a given global if used inside checks like:

```js
if (window.someGlobal) { ... }
if (typeof someGlobal) { ... }
```

Obviously there are a lot of ways to test for a global and it would probably be unreasonable for ESLint to support all of them, but I’d like to think this could still be useful. For one thing, not declaring a platform-specific global in configuration helps ensure that you always perform some sort of check before using it.

**Comment by isiahmeadows** [comment]
I'll note that this would be relatively difficult to fix, and I already brought it up in a somewhat more limited sense (exclusively `typeof`) in #6209.

**Comment by not-an-aardvark** [comment]
I think it would be reasonable to allow references to `foo` inside `if (typeof foo !== 'undefined') {}`.

Checking `if (window.foo) {}` is a bit more involved because we would have to make sure that `window` is the global object. For example, if someone was using Node and had a global variable `window` that was set to something other than the global object, `window.foo` is not actually checking for the existence of a variable called `foo`. It would also be possible to have weird things like a getter on the `window` object.

Personally, I think it would be best to just detect `typeof`, to encourage these patterns:

```js
if (typeof foo !== 'undefined') {
  foo();
}

if (window.foo) {
  window.foo();
}
```

`if (typeof foo !== 'undefined')` directly verifies that there is a variable called `foo`, so I think it makes sense for the rule to tolerate the use of `foo` in that block. `if (window.foo)` verifies something slightly different, so I don't think the rule should tolerate `foo` as a variable in that block (the user should use `window.foo` instead).

**Comment by Mr0grog** [comment]
> `if (typeof foo !== 'undefined')` directly verifies that there is a variable called `foo`, so I think it makes sense for the rule to tolerate the use of `foo` in that block.

I would at least say that, anecdotally, I see `if (window.foo)` *much* more than `if (typeof foo !== 'undefined')`, probably because the latter seems prohibitively verbose. That said, I certainly take your point on correctness. Actually abstractly verifying `if (somethingThatEqualsTheGlobalObject.foo)` is not easy (and not always reasonably possible in complex cases).

**Comment by platinumazure** [comment]
:+1: to typeof checks (only).

**Comment by isiahmeadows** [comment]
@platinumazure Should #6209 be re-opened, then? (That's exclusively `typeof`, nothing else.)

**Comment by Mr0grog** [comment]
This definitely seems like a dupe of #6209. I think I glossed over it when searching for similar issues because I assumed it was about the `typeof` support that already exists, not this. My mistake.

Has anything changed about how things are implemented in `no-undef` in the last few months? It sounds like the main reason #6209 was closed is that it was too complicated. I did a quick and dirty implementation of this earlier today (just the `typeof` check, not the `window.foo` method) and it seemed reasonably straightforward, but there could easily have been something I was missing.

**Comment by isiahmeadows** [comment]
@Mr0grog It was closed because it seemed complicated on the surface. If you've got a working implementation, go ahead and file a PR.

And yes, I believe this to be a dupe of that (which is why I brought it up in my first comment).

**Comment by isiahmeadows** [comment]
Do be aware of this:

1. There most certainly *should* be an `allowEscapingScopes` option, off by default, to allow the definitions to propagate into child function scopes, even though they theoretically be `delete`d from the global scope.

2. They *must* be counted as defined in block scopes, as well as their children, but not outside of them:

  ```js
  // These must not fail
  if (typeof foo !== "undefined") {
    while (cond()) {
      foo()
    }
    exports.foo = foo
  }

  // This must fail
  if (typeof foo !== "undefined") {
    bar()
  }

  foo()
  ```

3. They *must* be counted as defined in child IIFEs, provided there is no reference to `arguments.callee` or their name within them (excluding children), unless `allowEscapingScopes` is set.

  ```js
  // This should not fail
  if (typeof foo !== "undefined") {
    (function name() {
      (() => {
        foo
      })()
      foo
    })()
  }

  // This must fail
  if (typeof foo !== "undefined") {
    (function name() {
      foo
      array.map(name)
    })()
  }

  // This *may* fail
  if (typeof foo !== "undefined") {
    let count = 0
    (function loop() {
      foo
      if (++count < 5) loop()
    })()
  }
  ```

4. They *must not* be counted as defined in referenced child function scopes, unless `allowEscapingScopes` is set.

  ```js
  // These should all fail
  let method

  if (typeof foo !== "undefined") {
    exports.foo = () => foo
    method = function () { foo }
    return () => foo
    await () => foo
    map(() => foo)
  }
  ```

5. These rules are to be applied recursively as well.

The unsafe version is easy, but the safe one is a bit more difficult, and that's what you need to be aware of.

**Comment by nzakas** [comment]
Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.
